src/estimagic/optimization/subsolvers/_trsbox_quadratic_fast.py

```diff
@@ -0,0 +1,655 @@
+            else:
+                step_len_constr = (
+                    lower_bounds[i] - x_candidate[i]
+                ) / gradient_projected[i]
+            if step_len_constr < step_len:
+                step_len = step_len_constr
+                index_bound_active = np.array([i])
+
+    return step_len, index_bound_active
+
+
+@njit
+def _calc_upper_bound_on_tangent(
+    x_candidate,
+    search_direction,
+    x_bounded,
+    lower_bounds,
+    upper_bounds,
+    n_fixed_variables,
+):
+    """Calculate upper bound on tangent of half the angle to the boundary."""
+    bound_on_tangent = 1
+    free_variable_reached_bound = False
+
+    for i in range(len(x_candidate)):
+        if x_bounded[i] == 0:
+            lower_bound_centered = x_candidate[i] - lower_bounds[i]
+            upper_bound_centered = upper_bounds[i] - x_candidate[i]
+
+            if lower_bound_centered <= 0.0:
+                n_fixed_variables += 1
+                x_bounded[i] = -1
+                free_variable_reached_bound = True
+                break
+
+            elif upper_bound_centered <= 0.0:
+                n_fixed_variables += 1
+                x_bounded[i] = 1
+                free_variable_reached_bound = True
+                break
+
+            ssq = x_candidate[i] ** 2 + search_direction[i] ** 2
+
+            ssq_lower = ssq - lower_bounds[i] ** 2
+            if ssq_lower > 0.0:
+                ssq_lower = np.sqrt(ssq_lower) - search_direction[i]
+                if bound_on_tangent * ssq_lower > lower_bound_centered:
+                    bound_on_tangent = lower_bound_centered / ssq_lower
+                    index_active_bound = np.array([i])
+                    active_bound = np.array([-1])
+
+            ssq_upper = ssq - upper_bounds[i] ** 2
+            if ssq_upper > 0.0:
+                ssq_upper = np.sqrt(ssq_upper) + search_direction[i]
+                if bound_on_tangent * ssq_upper > upper_bound_centered:
+                    bound_on_tangent = upper_bound_centered / ssq_upper
+                    index_active_bound = np.array([i])
+                    active_bound = np.array([1])
+
+    return (
+        x_bounded,
+        index_active_bound,
+        n_fixed_variables,
+        active_bound,
+        bound_on_tangent,
+        free_variable_reached_bound,
+    )
+
+
+@njit
+def _calc_greatest_criterion_reduction(
+    bound_on_tangent, s_hess_s, x_hess_s, x_hess_x, x_grad, s_norm
+):
+    """Calculate the greatest feasible reduction in the criterion function.
+
+    The largest reduction is found by looking at a range of equally spaced values
+    of ``tangent`` in the interval [0, ``bound_on_tangent``], where ``tangent`` is
+    the tangent of half the angle to the trust-region boundary.
+    """
+    previous_reduction = None
+    next_reduction = None
+
+    max_reduction = 0
+    index_angle_greatest_reduction = -1
+    old_reduction = 0
+    n_angles = int(17 * bound_on_tangent + 3.1)
```

Comment:
Don't we need `np.int64` instead of just int? @timmens is this similar to the case you had?

Reply:
I think there is a difference between the `int()` method and passing an unsupported dtype to, for example, a `numpy.ndarray`. But maybe someone should take a look at the detailed log.

Otherwise, this could be done with `np.floor(...).astype(np.int64)`.